editor/gui/scene_tree_editor.cpp

```diff
@@ -1012,32 +1017,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
 	}
 }
 
-void SceneTreeEditor::_renamed() {
+void SceneTreeEditor::_edited() {
 	TreeItem *which = tree->get_edited();
-
 	ERR_FAIL_COND(!which);
-	NodePath np = which->get_metadata(0);
-	Node *n = get_node(np);
-	ERR_FAIL_COND(!n);
 
-	String new_name = which->get_text(0);
+	if (is_scene_tree_dock && tree->get_next_selected(which)) {
+		List<Node *> to_rename;
+		for (TreeItem *item = which; item; item = tree->get_next_selected(item)) {
+			Node *n = get_node(item->get_metadata(0));
+			ERR_FAIL_NULL(n);
+			to_rename.push_back(n);
+		}
+		ERR_FAIL_COND(to_rename.is_empty());
+
+		EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+		undo_redo->create_action(TTR("Rename Nodes"), UndoRedo::MERGE_DISABLE, to_rename.front()->get(), true);
+
+		TreeItem *item = which;
+		for (Node *n : to_rename) {
+			_renamed(item, which, n);
+			item = tree->get_next_selected(item);
+		}
+		which->remove_meta("use_class");
+		undo_redo->commit_action();
+	} else {
+		_renamed(which, nullptr);
+	}
+}
+
+void SceneTreeEditor::_renamed(TreeItem *p_item, TreeItem *p_batch_item, Node *p_node) {
+	Node *n;
+	if (p_node) {
+		n = p_node; // During batch rename the paths may change, so using metadata is unreliable.
+	} else {
+		n = get_node(p_item->get_metadata(0));
+	}
+	ERR_FAIL_NULL(n);
+
+	String new_name;
+	if (p_batch_item) {
+		if (!p_batch_item->get_meta(SNAME("use_class"), false)) {
+			new_name = p_batch_item->get_text(0);
+		}
+	} else {
+		new_name = p_item->get_text(0);
+	}
+
 	if (new_name.strip_edges().is_empty()) {
 		// If name is empty, fallback to class name.
 		if (GLOBAL_GET("editor/naming/node_name_casing").operator int() != NAME_CASING_PASCAL_CASE) {
 			new_name = Node::adjust_name_casing(n->get_class());
 		} else {
 			new_name = n->get_class();
 		}
+
+		// When doing batch rename, rename all nodes to their respective class.
+		if (p_batch_item == p_item) {
+			p_batch_item->set_meta("use_class", true);
+		}
 	}
 
 	if (n->is_unique_name_in_owner() && get_tree()->get_edited_scene_root()->get_node_or_null("%" + new_name) != nullptr) {
 		error->set_text(TTR("Another node already uses this unique name in the scene."));
 		error->popup_centered();
-		which->set_text(0, n->get_name());
+		p_item->set_text(0, n->get_name());
 		return;
 	}
```

Comment:
While looking at this I realized it was a problem they weren't inside `rename_node` so I tried a bit to see if I could things broke on master because of those and effectively I can :( So opening an issue for each ( #78290 
#78291) and a fix PR #78292